x/auth/middleware/basic.go

```diff
@@ -0,0 +1,359 @@
+package middleware
+
+import (
+	"context"
+
+	"github.com/cosmos/cosmos-sdk/codec/legacy"
+	"github.com/cosmos/cosmos-sdk/crypto/keys/multisig"
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	txtypes "github.com/cosmos/cosmos-sdk/types/tx"
+	"github.com/cosmos/cosmos-sdk/types/tx/signing"
+	"github.com/cosmos/cosmos-sdk/x/auth/migrations/legacytx"
+	authsigning "github.com/cosmos/cosmos-sdk/x/auth/signing"
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+// ValidateBasicDecorator will call tx.ValidateBasic, msg.ValidateBasic(for each msg inside tx)
+// and return any non-nil error.
+// If ValidateBasic passes, middleware calls next middleware in chain. Note,
+// validateBasicMiddleware will not get executed on ReCheckTx since it
+// is not dependent on application state.
+type validateBasicMiddleware struct {
+	next txtypes.Handler
+}
+
+func ValidateBasicMiddleware(txh txtypes.Handler) txtypes.Handler {
+	return validateBasicMiddleware{
+		next: txh,
+	}
+}
+
+var _ txtypes.Handler = validateBasicMiddleware{}
+
+// CheckTx implements tx.Handler.CheckTx.
+func (basic validateBasicMiddleware) CheckTx(ctx context.Context, tx sdk.Tx, req abci.RequestCheckTx) (abci.ResponseCheckTx, error) {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+
+	// no need to validate basic on recheck tx, call next antehandler
+	if sdkCtx.IsReCheckTx() {
```

Comment:
let's not use `sdkCtx.IsReCheckTx()`, but prefer `req.Type`, it'll be easier when we refactor out of sdkCtx in the future.